code/modules/vehicles/mecha/mecha_defense.dm

```diff
@@ -0,0 +1,324 @@
+	playsound(loc, 'sound/weapons/tap.ogg', 40, TRUE, -1)
+	user.visible_message(span_danger("[user] hits [src]. Nothing happens."), null, null, COMBAT_MESSAGE_RANGE)
+	log_message("Attack by hand/paw (no damage). Attacker - [user].", LOG_MECHA, color="red")
+
+/obj/vehicle/sealed/mecha/bullet_act(obj/projectile/hitting_projectile, def_zone, piercing_hit) //wrapper
+	log_message("Hit by projectile. Type: [hitting_projectile]([hitting_projectile.ammo.damage_type]).", LOG_MECHA, color="red")
+	// yes we *have* to run the armor calc proc here I love tg projectile code too
+	try_damage_component(run_obj_armor(hitting_projectile.damage, hitting_projectile.ammo.damage_type, hitting_projectile.ammo.armor_type, 0, REVERSE_DIR(hitting_projectile.dir), hitting_projectile.ammo.penetration), hitting_projectile.def_zone)
+	return ..()
+
+/obj/vehicle/sealed/mecha/ex_act(severity, target)
+	log_message("Affected by explosion of severity: [severity].", LOG_MECHA, color="red")
+	return ..()
+
+/obj/vehicle/sealed/mecha/contents_explosion(severity, target)
+	severity--
+
+	switch(severity)
+		if(EXPLODE_DEVASTATE)
+			if(flat_equipment)
+				SSexplosions.highMovAtom += flat_equipment
+			if(trackers)
+				SSexplosions.highMovAtom += trackers
+		if(EXPLODE_HEAVY)
+			if(flat_equipment)
+				SSexplosions.medMovAtom += flat_equipment
+			if(trackers)
+				SSexplosions.medMovAtom += trackers
+		if(EXPLODE_LIGHT)
+			if(flat_equipment)
+				SSexplosions.lowMovAtom += flat_equipment
+			if(trackers)
+				SSexplosions.lowMovAtom += trackers
+
+/obj/vehicle/sealed/mecha/handle_atom_del(atom/A)
+	if(A in occupants) //todo does not work and in wrong file
+		LAZYREMOVE(occupants, A)
+		icon_state = initial(icon_state)+"-open"
+		setDir(dir_in)
+
+/obj/vehicle/sealed/mecha/emp_act(severity)
+	. = ..()
+	if(get_charge())
+		use_power((cell.charge/3)/(severity*2))
+		take_damage(30 / severity, BURN, ENERGY, 1)
+	log_message("EMP detected", LOG_MECHA, color="red")
+
+	if(!equipment_disabled && LAZYLEN(occupants)) //prevent spamming this message with back-to-back EMPs
+		to_chat(occupants, span_warning("Error -- Connection to equipment control unit has been lost."))
+	addtimer(CALLBACK(src, /obj/vehicle/sealed/mecha.proc/restore_equipment), 3 SECONDS, TIMER_UNIQUE | TIMER_OVERRIDE)
+	equipment_disabled = TRUE
+	set_mouse_pointer()
+
+/obj/vehicle/sealed/mecha/fire_act() //Check if we should ignite the pilot of an open-canopy mech
+	. = ..()
+	if(enclosed || mecha_flags & SILICON_PILOT)
+		return
+	for(var/mob/living/cookedalive as anything in occupants)
+		if(cookedalive.fire_stacks < 5)
+			cookedalive.adjust_fire_stacks(1)
+			cookedalive.IgniteMob()
+
+/obj/vehicle/sealed/mecha/attackby_alternate(obj/item/weapon, mob/user, params)
+	if(istype(weapon, /obj/item/mecha_parts))
+		var/obj/item/mecha_parts/parts = weapon
+		parts.try_attach_part(user, src, TRUE)
+		return TRUE
+	return ..()
+
+/obj/vehicle/sealed/mecha/attackby(obj/item/W, mob/user, params)
```

Comment:
will mechs be constructable ?

Reply:
They shouldnt be if you mean whats happening here its attachables and swapping out upgradable parts